ast/src/functions/function.rs

```diff
@@ -60,6 +60,32 @@ impl Function {
         &self.identifier.name
     }
 
+    ///
+    /// Returns `true` if the function has input `self` or `mut self`.
+    /// Returns `false` otherwise.
+    ///
+    pub fn contains_self(&self) -> bool {
+        self.input.iter().any(|param| param.is_self())
+    }
+
+    ///
+    /// Returns `true` if the function has input `mut self`.
+    /// Returns `false` otherwise.
+    ///
+    pub fn contains_mut_self(&self) -> bool {
+        self.input.iter().any(|param| param.is_mut_self())
+    }
+
+    ///
+    /// Returns a vector of [&FunctionInput] removing `self` and `mut self` inputs.
+    ///
+    pub fn filter_self_inputs(&self) -> Vec<&FunctionInput> {
```

Comment:
it would be preferable to return `impl Iterator<Item = &FunctionInput>` here if possible; the only call I currently see immediately does `.iter()` on it, and even if this needs to be immediately collected somewhere, it would be better to do it at the callsite